botorch/utils/sample_pareto.py

```diff
@@ -0,0 +1,281 @@
+            )
+
+    if num_greedy > 0:
+        if X_baseline is None or ref_point is None:
+            raise UnsupportedError(
+                "Need to specify the `X_baseline` and `ref_point` in order to take "
+                "advantage of greedy truncation strategy."
+            )
+
+    pareto_sets = torch.zeros((num_pareto_samples, num_pareto_points, d), **tkwargs)
+    pareto_fronts = torch.zeros((num_pareto_samples, num_pareto_points, M), **tkwargs)
+
+    for i in range(num_pareto_samples):
+        model_sample_i = get_gp_samples(
+            model=model, num_outputs=M, n_samples=1, num_rff_features=num_rff_features
+        )
+
+        ratio = 2
+        num_tries = 0
+
+        # Run solver until we find at least `num_pareto_samples` optimal points
+        # or if the maximum number of retries is exceeded.
+        while (ratio > 1) and (num_tries < max_tries):
+            pareto_set_i, pareto_front_i = pareto_solver(
+                model=model_sample_i,
+                bounds=bounds,
+                num_objectives=M,
+                num_generations=num_generations,
+                pop_size=pop_size,
+                num_offsprings=num_offsprings,
+                maximize=maximize,
+            )
+            num_pareto_generated = pareto_set_i.shape[0]
+            ratio = ceil(num_pareto_points / num_pareto_generated)
+            num_tries = num_tries + 1
+
+        # If maximum number of retries exceeded throw out a runtime error.
+        if ratio > 1:
+            error_text = (
+                "Only found "
+                + str(num_pareto_generated)
+                + " Pareto efficient points instead of "
+                + str(num_pareto_points)
+                + "."
+            )
+            raise RuntimeError(error_text)
+
+        # Randomly truncate the Pareto set and front
+        if num_greedy == 0:
+            indices = torch.randperm(num_pareto_generated)[:num_pareto_points]
+
+        # Truncate Pareto set and front based on the sample hypervolume improvement
+        # or else select the points randomly.
+        else:
+            # get `num_pareto_points - num_greedy` indices randomly
+            num_remaining = max(0, num_pareto_points - num_greedy)
+            indices = torch.randperm(num_pareto_generated)[:num_remaining].tolist()
+
+            pending_pareto_front_i = model_sample_i.posterior(X_baseline).mean
+
+            for k in range(num_pareto_points - num_remaining):
+                partitioning = FastNondominatedPartitioning(
+                    ref_point=ref_point, Y=pending_pareto_front_i
+                )
+                hypercell_bounds = partitioning.hypercell_bounds
+
+                # `1 x num_boxes x M`
+                lo = hypercell_bounds[0].unsqueeze(0)
+                up = hypercell_bounds[1].unsqueeze(0)
+
+                # Compute the sample hypervolume improvement.
+                hvi = (
```

Comment:
It seems like this would be handy to have as an instance method of `BoxDecomposition`s (e.g. a method to compute the incremental HVI of a new point --- or more generally `(batch_shape x) 1 x M` new points)